File: Python/Double-define/rdf-nomath.py
import ROOT

LOOP_REPS=500000
N_ENTRIES=10000

verbosity = ROOT.Experimental.RLogScopedVerbosity(ROOT.Detail.RDF.RDFLogChannel(), ROOT.Experimental.ELogLevel.kInfo)

loop = f"int volatile s = 0; for (int i = 0; i < {LOOP_REPS}; ++i){{s += i;}}"

# Spaces are included to make the two strings different, and obtain two different jitted functions.
# If the strings match, the same function will appear in the flame graph
to_be_jitted_calc_x = f" {loop}; return (int)rdfentry_;"
to_be_jitted_calc_y = f"{loop}; return (int)rdfentry_; "

to_be_jitted_sum = f"{loop}; return x+y;"

def use_RDF():
	d = ROOT.RDataFrame(N_ENTRIES)
	dx = d.Define("x", to_be_jitted_calc_x)
	dy = dx.Define("y", to_be_jitted_calc_y)
	dz = dy.Define("z", to_be_jitted_sum)
	c = dz.Sum("z")
	print(c.GetValue())
	
if __name__ == "__main__":

	use_RDF()

File: C++/Double-define/rdf-nomath.cpp
#include <ROOT/RLogger.hxx>
#include <ROOT/RDataFrame.hxx>

#include <iostream>

#define LOOP_REPS 500000
#define N_ENTRIES 10000

auto verbosity = ROOT::Experimental::RLogScopedVerbosity(ROOT::Detail::RDF::RDFLogChannel(), ROOT::Experimental::ELogLevel::kInfo);

void waste_time(void)
{
        int volatile s = 0;
	for (int i = 0; i < LOOP_REPS; ++i){s += i;}
}

int calc_x(ULong64_t entry)
{
        waste_time();
        return (int)entry;
}

int calc_y(ULong64_t entry)
{
        waste_time();
        return (int)entry;
}


void use_RDF(void)
{
	auto d = ROOT::RDataFrame(N_ENTRIES);
	auto dx = d.Define("x", calc_x, {"rdfentry_"});
	auto dy = dx.Define("y", calc_y, {"rdfentry_"});
	auto dz = dy.Define("z", [](int x, int y){waste_time(); return x+y;}, {"x","y"});
	auto s = dz.Sum<int>("z");
	
	std::cout << "Complicated result: " << s.GetValue() <<std::endl;
	std::cout << "I am Double Define :D" << std::endl;
}

int main(void)
{
	use_RDF();
}
